src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,372 @@
+use crate::stacked_borrows::{AccessKind, Item, Permission, SbTag, SbTagExtra};
+use rustc_data_structures::fx::FxHashSet;
+#[cfg(feature = "stack-cache")]
+use std::ops::Range;
+
+/// Exactly what cache size we should use is a difficult tradeoff. There will always be some
+/// workload which has a `SbTag` working set which exceeds the size of the cache, and ends up
+/// falling back to linear searches of the borrow stack very often.
+/// The cost of making this value too large is that the loop in `Stack::insert` which ensures the
+/// entries in the cache stay correct after an insert becomes expensive.
+#[cfg(feature = "stack-cache")]
+const CACHE_LEN: usize = 32;
+
+/// Extra per-location state.
+#[derive(Clone, Debug)]
+pub struct Stack {
+    /// Used *mostly* as a stack; never empty.
+    /// Invariants:
+    /// * Above a `SharedReadOnly` there can only be more `SharedReadOnly`.
+    /// * Except for `Untagged`, no tag occurs in the stack more than once.
+    borrows: Vec<Item>,
+    /// If this is `Some(id)`, then the actual current stack is unknown. This can happen when
+    /// wildcard pointers are used to access this location. What we do know is that `borrows` are at
+    /// the top of the stack, and below it are arbitrarily many items whose `tag` is strictly less
+    /// than `id`.
+    /// When the bottom is unknown, `borrows` always has a `SharedReadOnly` or `Unique` at the bottom;
+    /// we never have the unknown-to-known boundary in an SRW group.
+    unknown_bottom: Option<SbTag>,
+
+    /// A small LRU cache of searches of the borrow stack. This only caches accesses of `Tagged`,
+    /// because those are unique in the stack.
+    #[cfg(feature = "stack-cache")]
+    cache: StackCache,
+    /// On a read, we need to disable all `Unique` above the granting item. We can avoid most of
+    /// this scan by keeping track of the region of the borrow stack that may contain `Unique`s.
+    #[cfg(feature = "stack-cache")]
+    unique_range: Range<usize>,
+}
+
+/// A very small cache of searches of the borrow stack
+/// This maps tags to locations in the borrow stack. Any use of this still needs to do a
+/// probably-cold random access into the borrow stack to figure out what `Permission` an
+/// `SbTag` grants. We could avoid this by also storing the `Permission` in the cache, but
+/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
```

Comment:
```suggestion
/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
///
/// It may seem like maintaining this cache is a waste for small stacks, but
/// (a) iterating over small fixed-size arrays is super fast, and (b) empirically this helps *a lot*,
/// probably because runtime is dominated by large stacks.
/// arrays is super fast
```